pages/blog/_posts/vercel-cloudflare-wait-until.mdx

```diff
@@ -0,0 +1,126 @@
+---
+heading: When to use Vercel and Cloudflare's waitUntil utility
+subtitle: What is it, when to use it, and when not to use it
+image: /assets/blog/vercel-cloudflare-wait-until/featured-image.png
+date: 2024-05-16
+author: Dan Farrelly
+---
+
+Recently, [Vercel announced](https://vercel.com/changelog/waituntil-is-now-available-for-vercel-functions) that their `waitUntil` utility method is now available for all serverless functions. Now, this useful feature is available for both Vercel's Node.js and edge functions. A similar utility is also available for [Cloudflare Workers](https://developers.cloudflare.com/workers/runtime-apis/context/#waituntil).
+
+This is a very useful tool, but it's important to understand when to use it and when not to use it. In this post, I'll explain:
+
+## What does `waitUntil` do
+
+To understand what `waitUntil` does, we need to understand how serverless functions work. This also helps us understand why it exists.
+
+Serverless functions are designed to be [stateless](https://www.redhat.com/en/topics/cloud-native-apps/stateful-vs-stateless) and short-lived. Unlike traditional servers, serverless functions are not kept alive between requests. Instead, they are spun up on-demand to handle incoming requests and then shut down once the response is sent.
+
+With JavaScript serverless functions, you have to `await` all asynchronous operations to ensure that the function doesn't return before the asynchronous operation is complete. If you don't `await` an asynchronous operation, the serverless function might be shut down before the operation is complete. If this happens, the operation might be canceled or fail which results in unpredictable behavior. This is why it's important to `await` all asynchronous operations in serverless functions.
+
+```js
+// ❌ BAD: If sendMetrics is an async function (a Promise), there is no guarantee it will succeed
+async function POST(request) {
+  const data = await createPost(request.body)
+  sendMetrics({ metric: 'posts.created', increment: 1 })
+  return new Response(data)
+}
+// ✅ GOOD: Awaiting will ensure it completes before the function returns
+async function POST(request) {
+  const data = await createPost(request.body)
+  await sendMetrics({ metric: 'posts.created', increment: 1 })
+  return new Response(data)
+}
+```
+
+The downside of this is that it can slow down the response time of your serverless functions. You ideally want to [keep the critical path of your request minimal](https://www.epicweb.dev/talks/improve-performance-and-reliability-of-your-api-with-events-and-background-functions) so you return the fastest response possible for your users. This is where `waitUntil` comes in.
```

Comment:
```suggestion
The downside of using `await` is that it can slow down the response time of your serverless functions. You ideally want to [keep the critical path of your request minimal](https://www.epicweb.dev/talks/improve-performance-and-reliability-of-your-api-with-events-and-background-functions) so you return the fastest response possible for your users. This is where `waitUntil` comes in.
```